### PR TITLE
SPU Mark II support

### PR DESCRIPTION
**Pull request by pixelherodev**
- [x] 16-bit ELF support
- [x] Basic linker scripts (specify RAM / ROM ranges)
  - [x] Use RAM and ROM correctly
- [x] Basic inline asm
- [x] Basic codegen
- [x] CI passing
- [ ] Add test
  - [ ] Add IHEX backend (separate PR), integrate with test harness and interpreter
  - [x] Add interpreter

Full support can come later, this is what's needed to ensure we can adequately target it.

Notable aspects of the SPU Mark II that make it worth targeting early to prove the design:

* Fully predicated. Every single instruction is predicated, so generating efficient code requires a different approach than for x86.
* 16-bit ISA. Currently, we're only supporting x64, so adding another bit width helps test parts of the design that are not yet set in stone.
* Custom linker scripts. e.g. AVR support will require something similar; we need to specify different types of memory and the ranges they occupy.

### CODE REVIEW

**Comment by pixelherodev** [comment]
Still a *lot* to go before this is ready to merge, this is just to track progress.

**Comment by pixelherodev** [comment]
Gah, needs a rebase.

**Comment by pixelherodev** [comment]
Working test program:

```zig
fn killEmulator() noreturn {
    // Probably kills the emulator, but no guarantee, so loop forever if it
    // doesn't.
    asm volatile ("undefined0");
}

export fn _start() noreturn {
    killEmulator();
}
```

After manually converting to IHEX, this results in 

```
[MAIN] Initializing system bus and attaching ROM...
[IHEX] load 0+16: 081E0280080020756E646566696E6564
[IHEX] load 16+2: 3000
[IHEX] load 32768+8: 0000050007000000
[MAIN] done!
[MAIN] Initializing CPU...
Executing instruction 0x1e08      imm  zero load16    jmp             imm0=0x8002                at 0x00000000
Executing instruction 0x2000      zero zero undefined discard                                    at 0x00000005
[MAIN] Emulation complete!
```

@MasterQ32

**Comment by FireFox317** [comment]
How is the killEmulator function noreturn? I think there is a unreachable missing after the assembly, because the compiler doesn't know that the assembly doesn't return right?

**Comment by pixelherodev** [comment]
... wat.

Before:

```
 ⚙ noam@pixelpc  ~/Documents/Development/osjw/zig/build   master ●  ./zig test ../test/stage1/behavior.zig --test-name-prefix behavior-wasm32-wasi-Debug-bare-single  --single-threaded --cache-dir ../zig-cache --name test -target wasm32-wasi -I ../test --override-lib-dir ../lib

LLVM Emit Output...LLVM Target Machine: wasm32
```

After:

```
 ⚙ noam@pixelpc  ~/Documents/Development/osjw/zig/build   spu_ii ●  ./zig test ../test/stage1/behavior.zig --test-name-prefix behavior-wasm32-wasi-Debug-bare-single  --single-threaded --cache-dir ../zig-cache --name test -target wasm32-wasi -I ../test --override-lib-dir ../lib

LLVM Emit Output...LLVM Target Machine: wasm64
LLVM ERROR: 64-bit WebAssembly (wasm64) is not currently supported
```

File: lib/std/spu/interpreter.zig [hunk 1]
```diff
@@ -0,0 +1,232 @@
+const std = @import("std");
+usingnamespace @import("defines.zig");
+
+const Bus = struct {
+    ROM: []u8,
+    RAM0: []u8,
+    RAM1: []u8,
```
Comment (MasterQ32) [inline lib/std/spu/interpreter.zig:7]: Is there a reason to separate RAM and ROM into seperate parts? It's not really necessary for testing the compiler backend, it's a hardware detail for Ashet, nothing that should matter for a generic backend

Comment (pixelherodev) [reply lib/std/spu/interpreter.zig:7]: Good point, I'll probably just give it a full address space of RAM instead or something like that.

**Comment by pixelherodev** [comment]
Okay, this currently depends on #5970, #5971, and #5972.

**Comment by andrewrk** [comment]
Everyone is excited about this, please upstream this in smaller mergeable chunks :)

**Comment by pixelherodev** [comment]
You *literally* just reviewed two of those PRs! :)

-- Email domain proudly hosted at https://migadu.com

**Pull request closed without merging**